File: backend/Testing/Browser/Component/BaseComponent.cs
using Microsoft.Playwright;

namespace Testing.Browser.Component;

public abstract class BaseComponent<T> where T : BaseComponent<T>
{
    public IPage Page { get; private set; }
    public ILocator ComponentLocator { get; }

    public BaseComponent(IPage page, ILocator componentLocator)
    {
        Page = page;
        ComponentLocator = componentLocator;
    }

    public virtual async Task<T> WaitFor()
    {
        await ComponentLocator.WaitForAsync();
        return (T)this;
    }

    public ILocator Locator(string selector, LocatorLocatorOptions? options = null)
    {
        return ComponentLocator.Locator(selector, options);
    }

    public virtual async Task AssertGone()
    {
        await Assertions.Expect(ComponentLocator).Not.ToBeAttachedAsync();
    }
}

File: frontend/tests/components/baseComponent.ts
import { expect, type Locator, type Page } from '@playwright/test';

export class BaseComponent {
  readonly page: Page;
  readonly componentLocator: Locator;

  constructor(page: Page, locator: Locator) {
    this.page = page;
    this.componentLocator = locator;
  }

  async waitFor(): Promise<this> {
    await this.componentLocator.waitFor();
    return this;
  }

  async assertGone(): Promise<void> {
    await expect(this.componentLocator).not.toBeAttached();
  }
}
